### PR TITLE
Add new DB merging API to distributor BucketDatabase 

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review

Abstracts away how an ordered merge may be performed with the database
and an arbitrary sorted bucket sequence, with any number of buckets
skipped, updated or inserted as part of the merge.

Such an API is required to allow efficient bulk updates of a B-tree
backed database, as it is suboptimal to require constant tree mutations.

Other changes:
- Removed legacy mutable iteration API. Not needed with new merge API.
- Const-iteration of bucket database now uses an explicit const reference
  entry type to avoid needing to construct a temporary entry when we can
  instead just point directly into the backing ArrayStore.
- Micro-optimizations of node remover pass to avoid going via cluster
  state's node state `std::map` for each bucket replica entry. Now uses
  a precomputed bit vector. Also avoid BucketId bit reversing operations
  as much as possible by using raw bucket keys in more places.
- Changed wording and contents of log message that triggers when buckets
  are removed from the DB due to no remaining nodes containing replicas
  for the bucket. Now more obvious what the message actually means.
- Added several benchmark tests (disabled by default)

### CODE REVIEW

File: storage/src/vespa/storage/bucketdb/bucketdatabase.h [hunk 1]
```diff
@@ -76,9 +82,37 @@ class BucketDatabase : public vespalib::Printable
...
-    virtual void forEach(
-            MutableEntryProcessor&,
-            const document::BucketId& after = document::BucketId()) = 0;
+
+    struct TrailingInserter {
+        virtual ~TrailingInserter() = default;
+        virtual void insert_at_end(const Entry&) = 0;
+    };
+
+    struct Merger {
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/bucketdatabase.h]: Consider adding class comment on how this is used.

File: storage/src/vespa/storage/bucketdb/bucketdatabase.h [hunk 2]
```diff
@@ -76,9 +82,37 @@ class BucketDatabase : public vespalib::Printable
...
+        // is _not_ returned is undefined.
+        // TODO this should ideally be separated into read/write functions, but this
+        // will suffice for now to avoid too many changes.
+        virtual uint64_t bucket_key() const noexcept = 0;
+        virtual document::BucketId bucket_id() const noexcept = 0;
+        virtual Entry& current_entry() = 0;
+        virtual void insert_before_current(const Entry&) = 0;
+    };
+
+    struct MergingProcessor {
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/bucketdatabase.h]: Consider adding class comment on how this is used in accordance with Merger.

File: storage/src/vespa/storage/bucketdb/bucketdatabase.h [hunk 3]
```diff
@@ -76,9 +82,37 @@ class BucketDatabase : public vespalib::Printable
...
+            KeepUnchanged,
+            Skip
+        };
+
+        virtual ~MergingProcessor() = default;
+        virtual Result merge(Merger&) = 0;
+        virtual void insert_remaining_at_end(TrailingInserter&) {}
+    };
+
+    virtual void merge(MergingProcessor&) = 0;
```
Comment (geirst) [inline storage/src/vespa/storage/bucketdb/bucketdatabase.h]: Consider adding comment on how an implementation of this is supposed to use the MergingProcessor interface.

**Review by geirst — APPROVED** [review]
👍 Consider adding some comments on how the new merge APIs are supposed to be used.

File: storage/src/vespa/storage/bucketdb/bucketdatabase.h (hunk 2 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/bucketdatabase.h]: Done

File: storage/src/vespa/storage/bucketdb/bucketdatabase.h (hunk 3 above)
Comment (vekterli) [reply storage/src/vespa/storage/bucketdb/bucketdatabase.h]: Done

**Comment by vekterli** [comment]
@geirst PTAL at added comments to see if they are fit for human consumption

**Review by geirst — APPROVED** [review]
Very good 🏅

**Pull request merged**